### PR TITLE
[core][state] Task state events backend for state API

### PR DESCRIPTION
**Pull request by rickyyx**
Signed-off-by: rickyyx <rickyx@anyscale.com>

## Why are these changes needed?
For details of the design and background see this [doc](https://docs.google.com/document/d/1l7GAbEoVbZszWOS2V1voSq5YiPzTgblK9dmtmnWb_tM/edit)

**TL;DR is we will emit task state changes from core workers/raylets to GCS in a similar way of how task profiling is currently done. This will allow persistence of task states to be queried by clients like State APIs or Dashboard.**

In this PR: 
1. Add a `TaskStateBuffer` which serves as an abstraction to store task state events, and push those events to GCS in batches. 
2. Add GCS gRPC server and client handler. 
3. Store the task state changes in GCS with capping for easy query. 
4. Make state APIs query GCS for task states. 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/core_worker/task_state_buffer.h [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+#include "ray/gcs/gcs_client/gcs_client.h"            // Gcs::GcsClient
+#include "src/ray/protobuf/gcs.pb.h"                  // rpc::TaskStateEventData
+
+namespace ray {
+namespace core {
+
+namespace worker {
+
+using TaskIdEventMap = absl::flat_hash_map<TaskID, rpc::TaskStateEvents>;
+/// An in-memory buffer for storing task state changes, and flushing them periodically to
```
Comment (rickyyx) [inline src/ray/core_worker/task_state_buffer.h:36]: This class is the major change.

**Comment by ericl** [comment]
We should also prototype a port the task timeline to this backend (the goal being to get performance parity switching from that profiler backend to this one).

File: src/ray/core_worker/task_state_buffer.cc [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+namespace ray {
+namespace core {
+
+namespace worker {
+
+TaskStateBuffer::TaskStateBuffer(instrumented_io_context &io_service,
+                                 const std::shared_ptr<gcs::GcsClient> &gcs_client)
+    : io_service_(io_service), periodical_runner_(io_service_), gcs_client_(gcs_client) {
+  periodical_runner_.RunFnPeriodically([this] { FlushEvents(); },
```
Comment (rkooo567) [inline src/ray/core_worker/task_state_buffer.cc:25]: Btw ,if you use `RunFnPeriodically `, the method can be called after this class is destructed. So you should make sure to handle that case within `FlushEvents`

File: src/ray/core_worker/task_state_buffer.h [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+  /// \param task_status Current task status to be recorded.
+  void AddTaskEvent(TaskID task_id, rpc::TaskStatus task_status) LOCKS_EXCLUDED(mutex_);
+
+ private:
+  TaskIdEventMap::iterator GetOrInitTaskEvents(TaskID task_id,
+                                               rpc::TaskStatus task_status)
+      EXCLUSIVE_LOCKS_REQUIRED(mutex_);
+
+  /// Flush all of the events that have been added since last flush to the GCS.
+  void FlushEvents() LOCKS_EXCLUDED(mutex_);
```
Comment (rkooo567) [inline src/ray/core_worker/task_state_buffer.h:76]: Make this public? We may need to do it before exiting worker in the future

File: src/ray/core_worker/task_state_buffer.cc [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+  task_info.set_task_id(task_spec.TaskId().Binary());
+  task_info.set_parent_task_id(task_spec.ParentTaskId().Binary());
+  const auto &resources_map = task_spec.GetRequiredResources().GetResourceMap();
+  task_info.mutable_required_resources()->insert(resources_map.begin(),
+                                                 resources_map.end());
+  task_info.mutable_runtime_env_info()->CopyFrom(task_spec.RuntimeEnvInfo());
+
+  // TODO(rickyx): Add node id later
+  // if (!node_id.IsNil()) {
```
Comment (rkooo567) [inline src/ray/core_worker/task_state_buffer.cc:96]: I guess we should have metadata & mutable data separation and this can be updated as a mutable field

File: src/ray/core_worker/task_state_buffer.cc (hunk 2 above)
Comment (rkooo567) [reply src/ray/core_worker/task_state_buffer.cc:25]: hmm maybe instead of periodically flushing it, we can flush upon reply? (with some delay). That could be a better backpressure policy? Doesn't need to be done from the prototype.

**Review by rkooo567 — COMMENTED** [review]
LGTM

File: src/ray/core_worker/task_state_buffer.cc (hunk 2 above)
Comment (rickyyx) [reply src/ray/core_worker/task_state_buffer.cc:25]: Ah that's good to know! Yeah, making it synchronous with reply could be a solution.

**Comment by rickyyx** [comment]
> We should also prototype a port of the task timeline to this backend (the goal being to get performance parity switching from that profiler backend to this one).

Yes, and I was thinking of developing this prototype independent of the codepaths related to task timeline with feature flag on, and refactor the task timeline part when the feature flag will be turned on by default

**Comment by ericl** [comment]
> Yes, and I was thinking of developing this prototype independent of the codepaths related to task timeline with feature flag on, and refactor the task timeline part when the feature flag will be turned on by default

So I think the goals of the prototype should be in this order:
1. Validate we can achieve feature parity with state API
2. Validate we can achieve feature parity with timeline API
3. Validate we achieve performance parity switched over to this new backend

**Comment by rickyyx** [comment]
> So I think the goals of the prototype should be in this order:
> 
> 1. Validate we can achieve feature parity with state API
> 2. Validate we can achieve feature parity with timeline API
> 3. Validate we achieve performance parity switched over to this new backend

Sounds good to me! Could I do this in maybe 3 PRs then? A single PR with all these might be rather huge 🤔

**Comment by ericl** [comment]
I think we can actually do all 3 at a rough level before we merge any code. Then, we can make multiple follow-up PRs to merge the feature assuming the design checks out. What do you think?

**Comment by rickyyx** [comment]
> I think we can actually do all 3 at a rough level before we merge any code. Then, we can make multiple follow-up PRs to merge the feature assuming the design checks out. What do you think?

This will also work I guess. I could always try to split them if this gets too large.

File: src/ray/gcs/gcs_server/gcs_task_manager.h [hunk 5]
```diff
@@ -0,0 +1,82 @@
+// Copyright 2022 The Ray Authors.
```
Comment (rickyyx) [inline src/ray/gcs/gcs_server/gcs_task_manager.h:1]: This is the major GCS side changes.

**Review by rickyyx — COMMENTED** [review]
Decided to rollback the profiling change but push minimal changes to be able to get a sense of the added overheads. Running a few release tests to see the performance impact. 

I wonder if we should merge this first as it's getting rather large (we could turn this off or guard with flags to prevent any impact), and the profiling refactoring is non-trivial (some cpython code changes at various places and rather orthogonal to this PR, adding a few protobuf fields). 

A few more known TODOs: 
- mock interfaces 
- make sure all tests passing.

File: src/ray/core_worker/task_state_buffer.cc (hunk 4 above)
Comment (rickyyx) [reply src/ray/core_worker/task_state_buffer.cc]: but i think in terms of performance, there's minimal impact?

**Comment by ericl** [comment]
> Decided to rollback the profiling change but push minimal changes to be able to get a sense of the added overheads. Running a few release tests to see the performance impact.

I think the two most relevant microbenchmarks are "multi client tasks async" and "n:n actor calls async". You can run both of these using "ray microbenchmark". I would do this before trying release tests.

> I wonder if we should merge this first as it's getting rather large

I'm really against this until we have a fully functioning prototype and can review our design choices. There's no need to do stuff like add unit tests or even implement most of the features here btw--- it should be a bare bones / POC prototype.

**Comment by rkooo567** [comment]
Btw, if we'd also like to test how it behaves under stress & large scale, `many_tasks` or `stress_test_many_tasks` could be useful. Another idea is to just start a local driver that submits 100K tasks (this will have pending 100K - num_cpus tasks, but it should be good enough to test load on GCS)

**Comment by rickyyx** [comment]
> I'm really against this until we have a fully functioning prototype and can review our design choices. There's no need to do stuff like add unit tests or even implement most of the features here btw--- it should be a bare bones / POC prototype.

I guess what I am suggesting is to not porting the profiling feature in the prototype. But I am open to get that in the PR as well.

**Comment by ericl** [comment]
We don't need to fully port it, just add enough information that we can assess the performance overhead and be confident we can port it. For example, the profiling includes events such as storing task outputs, etc., which may be significant to performance impact.

**Comment by rickyyx** [comment]
## Some local benchmark testing results: 

- baseline: This is where the new task backend is turned off (no task state tracking + flushing to GCS) 
- default: This is when profiling events for timeline is turned off (no profiling events tracking + flushing to GCS) 
- no-flush: This is when profiling turned off, but tracking is done and buffered at workers but not flushing to GCS. 
- quick-flush: Flush as quickly as possible. 

|    | enable_timeline   |   task_state_events_report_interval_ms | _name       |   one_n_actor_async |   single_client_tasks_async |   n_n_actor_async |   multi_client_tasks_async |
|---:|:------------------|------------------:|:------------|--------------------:|----------------------------:|------------------:|---------------------------:|
|  0 | True              |                 0 | baseline    |            10880.2  |                     11816.6 |           22521.2 |                    18495.7 |
|  1 | False             |              1000 | default     |             9646.47 |                     10246.1 |           20704.4 |                    16668.6 |
|  2 | False             |            200000 | no-flush    |            11054.5  |                     12580.6 |           24022.8 |                    19986.9 |
|  3 | False             |               100 | quick-flush |             9647.71 |                     10742.4 |           21705.8 |                    16653.9 |


tl;dr:
- A slight (5% - 10%) degrade in performance locally (default vs baseline) 
- Flushing events seems to be the major factor here. (no-flush is actually better).

**Comment by rkooo567** [comment]
Is default == master? 

What's the comparison between master vs flush task events?

**Comment by rickyyx** [comment]
> Is default == master?
> 
> What's the comparison between master vs flush task events (periodically)?

Baseline is the master, default is the new behaviour after disabling timeline and turned on this new PR feature (the intended default after this PR merged)

**Comment by rickyyx** [comment]
Breakdown of smaller PRs in  https://github.com/ray-project/ray/pull/30775 and others.
